### DISCUSSION TITLE
Change debugger type in DebugConfigurationProvider

### DISCUSSION THREAD

**Discussion by A2va** [microsoft/vscode-discussions#314] [Extension Development QnA] [answered]
I created an [issue](https://github.com/microsoft/vscode/issues/166862) in the vscode repository about this, if you want the full context.
To summarize, [xmake-vscode](https://github.com/xmake-io/xmake-vscode) invoke specific extensions (cpptools or codelldb) to debug a target. The problem is doesn't support a custom `launch.json` with custom arguments passing to the executable, for example.
So, my goal is to change the debugger type in the configuration provider.

An example of changing debug type from `xmake` to `cppdbg`.
```ts
class XmakeConfigurationProvider implements vscode.DebugConfigurationProvider {

	/**
	 * Massage a debug configuration just before a debug session is being launched,
	 * e.g. add all missing attributes to the debug configuration.
	 */
	resolveDebugConfiguration(folder: WorkspaceFolder | undefined, config: DebugConfiguration, token?: CancellationToken): ProviderResult<DebugConfiguration> {

		const programPath = getProgram(config.target) // Find the program from xmake target
			
		// Set the config type for platform specific or extension (lldb, cpp tools)
        config.type = 'cppdbg'; // or lldb
        config.name = 'Launch';
        config.request = 'launch';
        config.program = programPath;
        config.stopOnEntry = true;

		return config;
	}
}
```

On my issue roblourens think that it's already possible, but I didn't find anything about this.
Anyway I created a [mock extension](https://github.com/A2va/code-debug-conf-mock), this extension replace the `xmake`
debug type to `node` debug type and try to run a javascript file.

But it displays two error messages depending on the activation event (`onDebug` or `onDebugResolve`)
See comment: https://github.com/microsoft/vscode/issues/166862#issuecomment-1323782433

**Comment by bwateratmsft** [comment, marked as answer]
The Docker extension does something like this for debugging [Node](https://github.com/microsoft/vscode-docker/blob/main/src/debugging/node/NodeDebugHelper.ts), [.NET](https://github.com/microsoft/vscode-docker/blob/main/src/debugging/netcore/NetCoreDebugHelper.ts), and [Python](https://github.com/microsoft/vscode-docker/blob/main/src/debugging/python/PythonDebugHelper.ts). All three start out as a `"type": "docker"`, `"request": launch"` configuration and are resolved into one for their respective debuggers.

**Reply by A2va** [reply]
Thank you ! I managed it to work with python extension but not with node. I'm having this as a popup:
![image](https://user-images.githubusercontent.com/49582555/207433719-e25f10bc-029f-4aa1-9534-579fbb31fd22.png)
But it's not important I understood the concept so that's the main thing

**Reply by bwateratmsft** [reply]
Docker uses the type `pwa-node` for Node debugging. Truthfully, I don't remember why this specifically is chosen, but IIRC there are several options for Node debugging, and I think that `node` isn't one of them (ironically).
